In [1]:
import numpy as np
from statsmodels.stats.weightstats import ztest

В цьому ДЗ потренуємось виконувати Z-тести для тестування гіпотез в різних задачах. Для коректної імплементації та інтерпретації тестів рекоемндую переглянути спершу документацію:

https://www.statsmodels.org/dev/generated/statsmodels.stats.weightstats.ztest.html

### Завдання 1: Z-тест для однієї вибірки

**Мета**: Перевірити, чи середнє значення вибірки відрізняється від гіпотетичного середнього.

**Припущення**: Середній ріст студентів університету вважається 170 см. Ми хочемо перевірити, чи середній ріст випадкової вибірки студентів істотно більший. Вибірка задана в змінній `heights`.

Для виконання завдання імпортуйте метод для проведення `ztest`, виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи справдилось припущення про середній зріст студентів на рівні значущості 0.05?

Використовуємо тут Z-тест, оскільки вибірка достатньо велика (більше 30 екземплярів) і можемо вважати, що її стандартне відхилення близьке до ст. відхилення генеральної сукупності, а значить - відоме.

In [5]:
heights = [174, 171, 175, 179, 170, 170, 179, 175, 169, 174, 169, 169, 173, 162, 163, 169, 166,
           173, 167, 164, 179, 170, 172, 164, 169, 175, 169, 174, 169, 169, 173, 162, 177]

In [7]:
alpha = 0.05

In [9]:
z_stat, p_value = ztest(heights, value = 170, alternative='larger')
print(f"z_stat: {z_stat:.3f}")
print(f"p_value: {p_value:.3f}")

z_stat: 0.848
p_value: 0.198


In [11]:
if(p_value <  alpha):
    print("Відхилити Н0.")
else:
    print("Н0 не може бути відхилена.")

Н0 не може бути відхилена.


Висновок:
За результатами проведеного тестування, ми отримали: z_stat: 0.848, p_value: 0.198.
z-test проводився правосторонній, бо треба було визначити, чи середній ріст випадкової вибірки студентів істотно більший, тобто сер.знач. генеральної сукупності (вся популяція) > value(170) або >= value(170).
Так як z_stat = 0.848, ми бачимо, що не дуже далеко від гіпотетичного value(170).
p-value = 0.198 в нас ймовірність отримати таке або більше відхилення при умові, що нульова гіпотеза (Н0) вірна. Але в нас 0.198 > 0.05, тому висновок по p-value, такий що, Н0 не може бути відхилена.

### Завдання 2: Z-тест для двох незалежних вибірок

**Завдання 2.1.**

**Мета**: Перевірити, чи існує статистична різниця між середніми оцінками двох груп студентів.

**Припущення**: Група A отримала новий навчальний курс, а група B продовжувала зі стандартним курсом. Ми перевіряємо, чи є новий курс ефективнішим.

Виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи справдилось припущення щодо студентів на рівні значущості 0.05?

In [15]:
group_a_scores = [78.55, 72.25, 79.88, 75.  , 76.54, 74.99, 87.26, 77.93, 72.71,
       82.11, 71.9 , 79.04, 68.2 , 71.36, 78.98, 81.69, 78.86, 77.42,
       76.49, 70.61, 74.4 , 75.7 , 83.29, 79.72, 69.18, 79.62, 76.07,
       74.62, 81.06, 83.15, 82.66, 73.8 , 76.45, 79.66, 82.88, 75.6 ,
       77.07, 72.47, 72.02, 82.06]

group_b_scores = [81.78, 74.64, 80.02, 76.81, 71.77, 76.81, 82.69, 74.82, 82.82,
       61.9 , 79.11, 75.44, 73.5 , 75.46, 65.06, 73.9 , 76.79, 82.39,
       72.41, 70.96, 72.49, 79.58, 76.64, 72.35, 77.57, 75.49, 79.84,
       71.49, 73.36, 73.04, 67.68, 76.48, 76.31, 75.03, 73.83, 67.92,
       72.9 , 73.29, 70.99, 74.19]

In [17]:
np.mean(group_a_scores), np.mean(group_b_scores)

(77.08125, 74.73875000000001)

In [19]:
ztest(group_a_scores, group_b_scores, alternative='two-sided')

(2.357354864751502, 0.018405648552528013)

In [21]:
ztest(group_a_scores, group_b_scores, value = 0)

(2.357354864751502, 0.018405648552528013)

Висновок:
За результатами двусторонього проведеного z-тесту, ми отримали: z_test: 2.357, p_value: 0.0184.
Оскільки p-value (0.0184) < 0.05, у нас є достатньо доказів, щоб відкинути нульову гіпотезу. Іншими словами, ми можемо сказати, що новий курс є ефективнішим. 
Результати самого z_test: 2.357, кажуть нам про те, що різниця між групами є статистично значуща і не є випадковою.

**Завдання 2.2.**


**Мета**: Побачити, як впливає розмір вибірок на результат тесту.

**Завдання**: Уявіть, що з групи А ми маємо тільки перші 5 записів, а з групи В - усі. Таке могло статись, якщо ми вже маємо результати тестів учнів за попередньою програмою, а тести за новою програмою здали наразі лише 5 учнів і ми вирішили не чекати довше.
Виберіть перші 5 записів для групи А і проведіть t-test (для цього використовуємо наступний [метод](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html)). Виведіть значення p-value. Зробіть висновок, який результат ви бачите щодо стат. значущості на рівні 0.05 для цього експерименту?

Тут проводимо t-test оскільки одна з вибірок - дуже мала.

In [25]:
group_a_scores_new = [78.55, 72.25, 79.88, 75.  , 76.54]

In [27]:
np.mean(group_a_scores_new)

76.444

In [108]:
from scipy.stats import ttest_ind

t_stat, p_value = ttest_ind(group_a_scores_new, group_b_scores)

print(f"t-статистика: {t_stat}")
print(f"p-value: {p_value}")


t-статистика: 0.8167951030410814
p-value: 0.4185479477535733


In [31]:
if(p_value <  alpha):
    print("Відхилити Н0.")
else:
    print("Н0 не може бути відхилена.")

Н0 не може бути відхилена.


Висновок:
Оскільки p-value = 0.4185 > 0.05, немає статистично значущої різниці між середніми оцінками груп. Нульова гіпотеза не відхиляється.

Вплив на експеремент спостерігається такий, що зменшення вибірки повпливало на імовірність, що така різниця (було - p_value: 0.0184 / стало - p-value = 0.4185) могла виникнути випадково, стала великою. Як висновок, далі стверджувати, що новий курс став ефективніший ми не можемо. 


### Завдання 3\*: Z-тест для двох **пов'язаних** вибірок

**Мета**: Перевірити, чи вплинув тренінг на продуктивність працівників, коли всі справробітники проходили тренінг.

**Припущення**: Наші працівники проходять однаковий тренінг, і ми хочемо з'ясувати, чи покращилися їхні показники після тренінгу на рівні значущості 0.05. Виконайте Z-тест з Python, виведіть p-value і зробіть заключення, чи поліпшились показники працівників на рівні значущості 0.05?

Зверніть увагу, ці вибірки повʼязані між собою, значить не буде коректно виконати z-test між ними, натомість, ми можемо порівняти, чи різниця в результатах (продуктивності співробітників) є стат значущою.

In [35]:
before_training = [57.82, 37.63, 36.8 , 55.22, 52.97, 52.5 , 53.46, 43.2 , 52.32,
       52.93, 42.86, 68.66, 54.74, 38.09, 56.57, 40.25, 57.87, 61.59,
       41.79, 59.63, 54.13, 58.22, 68.97, 47.55, 42.46, 41.1 , 41.84,
       49.23, 53.41, 52.77]

after_training = [62.47, 40.66, 42.7 , 57.69, 61.41, 56.76, 54.75, 44.06, 56.29,
       55.48, 47.28, 72.6 , 57.59, 39.39, 56.54, 42.36, 62.58, 65.01,
       42.3 , 62.98, 57.9 , 59.45, 72.28, 50.66, 43.18, 44.82, 45.96,
       54.4 , 58.52, 53.01]

In [70]:
before_training_update = np.array([57.82, 37.63, 36.8 , 55.22, 52.97, 52.5 , 53.46, 43.2 , 52.32,
       52.93, 42.86, 68.66, 54.74, 38.09, 56.57, 40.25, 57.87, 61.59,
       41.79, 59.63, 54.13, 58.22, 68.97, 47.55, 42.46, 41.1 , 41.84,
       49.23, 53.41, 52.77])

In [72]:
after_training_update = np.array([62.47, 40.66, 42.7 , 57.69, 61.41, 56.76, 54.75, 44.06, 56.29,
       55.48, 47.28, 72.6 , 57.59, 39.39, 56.54, 42.36, 62.58, 65.01,
       42.3 , 62.98, 57.9 , 59.45, 72.28, 50.66, 43.18, 44.82, 45.96,
       54.4 , 58.52, 53.01])

In [83]:
diff = after_training_update - before_training_update
diff

array([ 4.65,  3.03,  5.9 ,  2.47,  8.44,  4.26,  1.29,  0.86,  3.97,
        2.55,  4.42,  3.94,  2.85,  1.3 , -0.03,  2.11,  4.71,  3.42,
        0.51,  3.35,  3.77,  1.23,  3.31,  3.11,  0.72,  3.72,  4.12,
        5.17,  5.11,  0.24])

Після виведення різниці згідно проведеного тестування переходжу до тестування, використовуючи 

In [85]:
mean_diff = np.mean(diff)
mean_diff

3.1499999999999995

In [96]:
std_diff = np.std(diff, ddof=0)
z_stat = mean_diff / (std_diff / np.sqrt(30))

In [112]:
from scipy.stats import norm
p_value = 1 - norm.cdf(z_stat) ##cdf — кумулятивна функція розподілу

In [114]:
z_stat, p_value

(9.29513192774473, 0.0)

Оскільки p-value (0.0) менше рівня значущості 0.05, ми відхиляємо нульову гіпотезу. Це свідчить про те, що після проходження тренінгу показники працівників статистично значущо покращилися. Можемо стверджувати, що є позитивний ефектк на продуктивність працівників.